scripts/source-maintenance.sh

```diff
@@ -340,15 +384,15 @@ for FILENAME in `git ls-files`; do
     esac
 
     # check for Foundation copyright blurb
-    if test $CheckAndUpdateCopyright = yes -a -f ${FILENAME} -a "x$skip_copyright_check" = "x"; then
-        BLURB=`grep -o "${COPYRIGHT_YEARS} The Squid Software Foundation and contributors" ${FILENAME}`;
-        if test "x${BLURB}" = "x"; then
-            BOILER=`grep -o -E "1996-2[0-9]+ The Squid Software Foundation and contributors" ${FILENAME}`;
-            if test "x${BOILER}" != "x"; then
-                echo "UPDATE COPYRIGHT for ${FILENAME}"
-                sed --in-place -r -f boilerplate_fix.sed ${FILENAME}
+    if test ${CheckAndUpdateCopyright} = yes -a -f ${FileName} -a "x$skip_copyright_check" = "x"; then
+        Blurb=`grep -o "${CopyRightYears} The Squid Software Foundation and contributors" ${FileName}`;
+        if test "x${Blurb}" = "x"; then
+            BoilerPlate=`grep -o -E "1996-2[0-9]+ The Squid Software Foundation and contributors" ${FileName}`;
+            if test "x${BoilerPlate}" != "x"; then
+                echo "UPDATE COPYRIGHT for ${FileName}"
+                sed --in-place -r -f boilerplate_fix.sed ${FileName}
             else
-                echo "CHECK COPYRIGHT for ${FILENAME}"
+                echo "CHECK COPYRIGHT for ${FileName}"
```

Comment:
IMO, this out-of-scope change and similar changes throughout the code should be undone. Some of them are desirable, some are wrong, but they should not be applied to otherwise unmodified-by-the-PR code because such widespread changes make review unnecessary difficult.

Reply:
Sure, I can do that, but your statement raises an interesting issue: who decides what the scope for a PR is?
Some of our discussions in the past have been (or at least have felt to me) disagreements on what is the scope for change. I believe we could improve the efficiency and effectiveness of how we work together if we could define this point, and how to deal with disagreements on it.
By the way, I would in general prefer smaller PRs with more targeted changes. Maintaining a branch long term is a hassle, and I have shown I'm not great at it (as a general statement, this case is clearly trivial)

Reply:
> who decides what the scope for a PR is?

The PR author decides what the PR scope is and describes that scope in the PR title/description. A reviewer may block the PR with a wrong/unclear (in reviewer opinion) scope or with a mismatch between the PR scope and code changes.

For example, the documented scope of this PR does not include renaming variables. The PR scope is not a problem, but renaming variables is (in this specific case). If you were to change the PR scope to also include variable renaming, then I would block the PR on "bad scope" grounds and ask you to split the PR into two: renaming variables and functionality changes (each of which would then be reviewed and accepted or blocked separately).

N.B. _If_ you do want to submit a variable renaming PR, let's agree on what naming/dereference style we should use for shell scripts first!

> Some of our discussions in the past have been (or at least have felt to me) disagreements on what is the scope for change

Yes, that is a common area of disagreements, for several reasons.

> I believe we could improve the efficiency and effectiveness of how we work together if we could define this point, and how to deal with disagreements on it.

Agreed. I always try to document the rationale behind my change requests (scope-driven or not), but sometimes I miss a case, especially if it feels obvious to me. If the (perceived) end result of the change request would move us backwards (in your opinion), please ask for clarifications. I make mistakes, but, statistically speaking, the vast majority of my change requests are asking for PR changes that should be done, but the vast majority of time is wasted on arguing about irrelevant things because the author is trying to defend their code. I have [documented](https://github.com/measurement-factory/squid-notes/blob/start/review.md) many aspects of my overall approach to reviews. As always, improvement suggestions welcome!

> By the way, I would in general prefer smaller PRs with more targeted changes. Maintaining a branch long term is a hassle, and I have shown I'm not great at it (as a general statement, this case is clearly trivial)

If you prefer smaller PRs, I see no reasons why you should not post smaller PRs! It is _your_ choice. Just make sure that the PR scope is reasonable. In most cases, a reasonable reviewer should not block a PR just because the scope is "small" or "large".

For example, this PR could have been split into three (or even more) PRs -- it adds several completely independent changes. The `-Wextra` PR could have been split into many PRs, each adding a single warning or two (with a final PR replacing all of them with the `-Wextra`).

On the other hand, a PR doing two completely unrelated/orthogonal things should usually be split, especially when one change is obscuring the other, making correct review impractical. This is one of the reasons why global-renaming PRs should be split from functionality-changing PRs.

Reply:
> > who decides what the scope for a PR is?
> 
> The PR author decides what the PR scope is and describes that scope in the PR title/description. A reviewer may block the PR with a wrong/unclear (in reviewer opinion) scope or with a mismatch between the PR scope and code changes.
> 
> For example, the documented scope of this PR does not include renaming variables. The PR scope is not a problem, but renaming variables is (in this specific case). If you were to change the PR scope to also include variable renaming, then I would block the PR on "bad scope" grounds and ask you to split the PR into two: renaming variables and functionality changes (each of which would then be reviewed and accepted or blocked separately).

Variables were renamed as a followup action to a [previous comment of yours](https://github.com/squid-cache/squid/pull/830#discussion_r638799054): if there is a readability problem with CHANGEDONLY, there is a much bigger one with ASVER, COPYRIGHT_YEARS, or FI . 
So I suppose the problem here is that I misinterpreted your request as a scope change request.

I'll amend the scope definition in the PR description; let's converge there.

> N.B. _If_ you do want to submit a variable renaming PR, let's agree on what naming/dereference style we should use for shell scripts first!

Agreed.
our biggest shell script is ./configure, which uses long_variable_names_with_underscores 

> > Some of our discussions in the past have been (or at least have felt to me) disagreements on what is the scope for change
> 
> Yes, that is a common area of disagreements, for several reasons.
> 
> > I believe we could improve the efficiency and effectiveness of how we work together if we could define this point, and how to deal with disagreements on it.
> 
> Agreed. I always try to document the rationale behind my change requests (scope-driven or not), but sometimes I miss a case, especially if it feels obvious to me. If the (perceived) end result of the change request would move us backwards (in your opinion), please ask for clarifications. I make mistakes, but, statistically speaking, the vast majority of my change requests are asking for PR changes that should be done, but the vast majority of time is wasted on arguing about irrelevant things because the author is trying to defend their code. I have [documented](https://github.com/measurement-factory/squid-notes/blob/start/review.md) many aspects of my overall approach to reviews. As always, improvement suggestions welcome!

Let's move the discussion to that note then, it will be more productive

> > By the way, I would in general prefer smaller PRs with more targeted changes. Maintaining a branch long term is a hassle, and I have shown I'm not great at it (as a general statement, this case is clearly trivial)
> 
> If you prefer smaller PRs, I see no reasons why you should not post smaller PRs! It is _your_ choice. Just make sure that the PR scope is reasonable. In most cases, a reasonable reviewer should not block a PR just because the scope is "small" or "large".

Okay then, I will then probably favour splitting large chunks of work into a string of PRs, and it may happen that scope change requests will be moved to immediate followup PRs - I don't want to leave things halfway through, of course.
Do we agree on this?

> For example, this PR could have been split into three (or even more) PRs -- it adds several completely independent changes. The `-Wextra` PR could have been split into many PRs, each adding a single warning or two (with a final PR replacing all of them with the `-Wextra`).

This, I agree.
Wextra, possibly, altthough I would have favoured an approach where:
- Wextra is added but default-off
- I attack each class of errors with an own PR (or multiple classes, if there are very few instances of errors), testing by adding Wextra in test builds
- after all are cleared, a PR switches Wextra default-on

This reflects the fact that for different compilers Wextra means different sets of Wsomething flags

> On the other hand, a PR doing two completely unrelated/orthogonal things should usually be split, especially when one change is obscuring the other, making correct review impractical. This is one of the reasons why global-renaming PRs should be split from functionality-changing PRs.

Agreed, in general. Given the well-constrained scope of this PR, I felt it was okay, but I'll undo it.

So, to recap:
- I'll undo the variables renaming
- I'll keep new added variables as CamelCase
- external variables (ASTYLE, CHECKSUM) will remain capitalised - they are environment
- I don't like the style of a CONFIG_ something variable, I will discard that suggestion
- I'll add a --with-checksum flag for consistency, if you ask for it - I don't expect it will ever be used
- I'll fix the debug-sections bug (by not updating debug-sections.txt if we don't scan all files)

Does this sound a good scope definition?

Reply:
> Variables were renamed as a followup action to a [previous comment of yours](https://github.com/squid-cache/squid/pull/830#discussion_r638799054): if there is a readability problem with CHANGEDONLY, there is a much bigger one with ASVER, COPYRIGHT_YEARS, or FI .

Agreed. Those problematic existing variables should not have been renamed (in this PR). The change request did not ask to rename them.

> So I suppose the problem here is that I misinterpreted your request as a scope change request.

I cannot be the judge of that.

FWIW, the change request was about a single new variable. The same approach should have been naturally applied to all new variables, and the followup discussion did mention "new variables". All those new variable names/changes do not change PR scope.

> I'll amend the scope definition in the PR description; let's converge there.

We still disagree regarding using environment variables instead of (or in addition to) configuration options.


> our biggest shell script is ./configure, which uses long_variable_names_with_underscores

./configure is a generated file. You probably meant its source -- ./configure.ac and friends. Those use a mixture of styles, including CamelCase and snake_case. Please note that variable names supplied by a 3rd party (e.g. `ac_cv_path_PERL`) do not really count in this popularity contest. While I personally think that it would be slightly better to have the same CamelCase style across C and shell sources, I am not against standardizing on snake_case if others prefer that alternative.
 
> Let's move the discussion to that note then, it will be more productive

I am not sure _how_ we can discuss a file in a repository -- create a fake pull request? We can discuss it on one of the mailing lists. Whatever works best for you. _If_ you and others think that many of those notes should be (after review/polishing) adopted by the Squid Project, then I can create the corresponding official empty repository and post a PR with the current start branch contents (or parts of it), which will allow everybody to review it. The creation of a new Project repository needs consensus, of course, even if the repository is going to be empty at first.

> Okay then, I will then probably favour splitting large chunks of work into a string of PRs, and it may happen that scope change requests will be moved to immediate followup PRs - I don't want to leave things halfway through, of course.
> Do we agree on this?

I am not quite sure what "scope change requests will be moved to immediate followup PRs" agreement would mean to you, so I hesitate answering "yes". To me, each PR should be evaluated on its own merits. If PR author wants to split their PR, and each remaining part still merits to be a stand-alone PR, then nobody should object to such splitting.

However, there are some PRs that are very difficult to evaluate correctly. For example, a PR that introduces a new API without sufficient code changes to use that API. IIRC, most of such PRs that I have seen had serious API design problems, but those problems were difficult to detect (and then convince the author that they existed) because there were no use cases for the API -- the problems were left unexposed. There are limits to this divide and conquer approach. We have to see where we are going to evaluate a step.


> Wextra, possibly, altthough I would have favoured an approach where:

> * Wextra is added but default-off

I see no value in this specific PR. One can already enable `-Wextra` without any Squid changes.
 
> * I attack each class of errors with an own PR (or multiple classes, if there are very few instances of errors),

Yes.

> ... testing by adding Wextra in test builds

Probably not -- `-Wextra` tests would fail (until the very last PR in the series). Instead, each "class of errors" PR would simply add `-Wfoo` option(s) that represent that specific class. Those tests would succeed if the PR covered all the cases in the PR-specific class.

Every PR/step would be a long-term improvement towards a well-understood worthy goal.

> * after all are cleared, a PR switches Wextra default-on

Yes, it will replace all those individual `-Wfoo` options(s) with one `-Wextra`.
 

> This reflects the fact that for different compilers Wextra means different sets of Wsomething flags

A class-at-a-time approach reflects that fact as well, but also does not add unusable options and guarantees that fixed official code (for class A) stays fixed while you work on classes B and C -- automated tests will detect and reject any new violations in already covered classes. Moreover, even if we give up in the middle of the sequence, we would still leave Squid in a good state, with all the already covered classes staying fixed forever and no unusable options added.

> > On the other hand, a PR doing two completely unrelated/orthogonal things should usually be split, especially when one change is obscuring the other, making correct review impractical. This is one of the reasons why global-renaming PRs should be split from functionality-changing PRs.
 
> Agreed, in general. Given the well-constrained scope of this PR, I felt it was okay, but I'll undo it.

It was not OK because I lost the ability to review your changes -- they became obscured by all the renaming going on around them.

> So, to recap:
> 
> * I'll undo the variables renaming

OK (especially as far as _old_ variables are concerned).
 
> * I'll keep new added variables as CamelCase

OK. If you think we should use standardize on snake_case instead, please discuss.
 
> * external variables (ASTYLE, CHECKSUM) will remain capitalised - they are environment

The external variables should use all-caps and, hence, SNAKE_CASE (where we control the new name). We may be in agreement here.

I do not think we should have external variables for ASTYLE and CHECKSUM features. That disagreement remains unresolved.
 
> * I don't like the style of a CONFIG_ something variable, I will discard that suggestion

Your call, but it was a way to give you what you wanted -- a variable in .bashrc -- without me objecting to duplicating individual command line options with environment variables. I do not see a better option to address your need without triggering my objections.

> * I'll add a --with-checksum flag for consistency, if you ask for it - I don't expect it will ever be used

I believe that consistency should be _preserved_ by _not_ adding support for ASTYLE and CHECKSUM environment variables.

BTW, why do you need `MD5`/`CHECKSUM` at all? What is the actual use case there, assuming the script is enhanced to find/guess a usable checksum program? Is your favorite checksum program not on the proposed guessing list?


> * I'll fix the debug-sections bug (by not updating debug-sections.txt if we don't scan all files)

This is a tricky problem: We should discuss whether it would be better to keep debug-sections.txt in sync (by scanning all files) instead of creating a special case and leaving debug-sections.txt stale (by not updating it at all). What do you think?

Reply:
> > Variables were renamed as a followup action to a [previous comment of yours](https://github.com/squid-cache/squid/pull/830#discussion_r638799054): if there is a readability problem with CHANGEDONLY, there is a much bigger one with ASVER, COPYRIGHT_YEARS, or FI .
> 
> Agreed. Those problematic existing variables should not have been renamed (in this PR). The change request did not ask to rename them.
> 
> > So I suppose the problem here is that I misinterpreted your request as a scope change request.
> 
> I cannot be the judge of that.
> 
> FWIW, the change request was about a single new variable. The same approach should have been naturally applied to all new variables, and the followup discussion did mention "new variables". All those new variable names/changes do not change PR scope.

TBH, this script is a right mess. I wouldn't be surprised if it broke if any file name contained a space.
I will address that in a followup PR

> > I'll amend the scope definition in the PR description; let's converge there.
> 
> We still disagree regarding using environment variables instead of (or in addition to) configuration options.

Yes

> 
> > our biggest shell script is ./configure, which uses long_variable_names_with_underscores
> 
> ./configure is a generated file. You probably meant its source -- ./configure.ac and friends. Those use a mixture of styles, including CamelCase and snake_case. Please note that variable names supplied by a 3rd party (e.g. `ac_cv_path_PERL`) do not really count in this popularity contest. While I personally think that it would be slightly better to have the same CamelCase style across C and shell sources, I am not against standardizing on snake_case if others prefer that alternative.

There is no reason outside personal preference to pick one over the other. I prefer snake_case, but in a popularity contest I would vote for CamelCase for consistency.

> > Okay then, I will then probably favour splitting large chunks of work into a string of PRs, and it may happen that scope change requests will be moved to immediate followup PRs - I don't want to leave things halfway through, of course.
> > Do we agree on this?
> 
> I am not quite sure what "scope change requests will be moved to immediate followup PRs" agreement would mean to you, so I hesitate answering "yes". To me, each PR should be evaluated on its own merits. If PR author wants to split their PR, and each remaining part still merits to be a stand-alone PR, then nobody should object to such splitting.

For instance, in the case of source-maintenance.sh I will address the proper quoting of variable names in a followup PR instead of taking on it here

> However, there are some PRs that are very difficult to evaluate correctly. For example, a PR that introduces a new API without sufficient code changes to use that API. IIRC, most of such PRs that I have seen had serious API design problems, but those problems were difficult to detect (and then convince the author that they existed) because there were no use cases for the API -- the problems were left unexposed. There are limits to this divide and conquer approach. We have to see where we are going to evaluate a step.

Agreed. What I would suggest in that approach is to have a PR with the additional API and one user, then a followup PR for each additional user, and finally a PR to remove the old API

> > Wextra, possibly, altthough I would have favoured an approach where:
> 
> > * Wextra is added but default-off
> 
> I see no value in this specific PR. One can already enable `-Wextra` without any Squid changes.

Fair point.

> > * I attack each class of errors with an own PR (or multiple classes, if there are very few instances of errors),
> 
> Yes.
> 
> > ... testing by adding Wextra in test builds
> 
> Probably not -- `-Wextra` tests would fail (until the very last PR in the series). Instead, each "class of errors" PR would simply add `-Wfoo` option(s) that represent that specific class. Those tests would succeed if the PR covered all the cases in the PR-specific class.

make -k and search in the logs for an error message would enable iterations just fine, wouldn't they?

> Every PR/step would be a long-term improvement towards a well-understood worthy goal.
> 
> > * after all are cleared, a PR switches Wextra default-on
> 
> Yes, it will replace all those individual `-Wfoo` options(s) with one `-Wextra`.
> 
> > This reflects the fact that for different compilers Wextra means different sets of Wsomething flags
> 
> A class-at-a-time approach reflects that fact as well, but also does not add unusable options and guarantees that fixed official code (for class A) stays fixed while you work on classes B and C -- automated tests will detect and reject any new violations in already covered classes. Moreover, even if we give up in the middle of the sequence, we would still leave Squid in a good state, with all the already covered classes staying fixed forever and no unusable options added.

Squid's state would be improved anyway, as addressing uncovered issues would stay. But I get your point on them staying fixed.

> > > On the other hand, a PR doing two completely unrelated/orthogonal things should usually be split, especially when one change is obscuring the other, making correct review impractical. This is one of the reasons why global-renaming PRs should be split from functionality-changing PRs.
> 
> > Agreed, in general. Given the well-constrained scope of this PR, I felt it was okay, but I'll undo it.
> 
> It was not OK because I lost the ability to review your changes -- they became obscured by all the renaming going on around them.
> 
> > So, to recap:
> > 
> > * I'll undo the variables renaming
> 
> OK (especially as far as _old_ variables are concerned).

Done
> 
> > * I'll keep new added variables as CamelCase
> 
> OK. If you think we should use standardize on snake_case instead, please discuss.

I prefer snake_case, but why having different styles for different languages? code is code

> > * external variables (ASTYLE, CHECKSUM) will remain capitalised - they are environment
> 
> The external variables should use all-caps and, hence, SNAKE_CASE (where we control the new name). We may be in agreement here.

Agreed

> I do not think we should have external variables for ASTYLE and CHECKSUM features. That disagreement remains unresolved.

Yes. I believe it follows best practices. Besides my previous example on ./configure (think CC, CXXFLAGS, etc), it is also used in most Unix software. A software using an environment variable to pass command line options is a white fly - I can't think of a single example. Whereas EDITOR is a de-facto standard, for example.

> > * I don't like the style of a CONFIG_ something variable, I will discard that suggestion
> 
> Your call, but it was a way to give you what you wanted -- a variable in .bashrc -- without me objecting to duplicating individual command line options with environment variables. I do not see a better option to address your need without triggering my objections.

I hope that the examples I brought will allow you to accept my point of view and let it pass

> > * I'll add a --with-checksum flag for consistency, if you ask for it - I don't expect it will ever be used
> 
> I believe that consistency should be _preserved_ by _not_ adding support for ASTYLE and CHECKSUM environment variables.
> 
> BTW, why do you need `MD5`/`CHECKSUM` at all? What is the actual use case there, assuming the script is enhanced to find/guess a usable checksum program? Is your favorite checksum program not on the proposed guessing list?

It is. We may think it's overengineering and decide not to do anything special about it, but the total code difference it makes is:
for CHECKSUM in "${CHECKSUM:-md5}" md5sum shasum sha1sum
versus
for CHECKSUM in md5 md5sum shasum sha1sum



> > * I'll fix the debug-sections bug (by not updating debug-sections.txt if we don't scan all files)
> 
> This is a tricky problem: We should discuss whether it would be better to keep debug-sections.txt in sync (by scanning all files) instead of creating a special case and leaving debug-sections.txt stale (by not updating it at all). What do you think?

Debug-sections.txt changes very rarely. Anyone adding a new debug section is likely to add it by hand, and it would be sorted correctly the next time the source-maintenance is run on a full sweep anyway.
On the other hand a full source scan would kill feasibility of the (arguably way more common) use case of a pre-commit hook to enforce source formatting on changed parts only, to make it fast

Reply:
> TBH, this script is a right mess.

Yes, it is. And so is a lot of Squid code.

> I wouldn't be surprised if it broke if any file name contained a space.

Yes, I bet it will break.

> I will address that in a followup PR

FWIW, we do not want to support spaces in git-controlled paths. However, "external" paths (computed or supplied by admins) may indeed contain spaces, especially on Windows, and we should support them.

I suspect we should switch from shell to Perl (or better) instead of spending a lot of time on perfecting a non-trivial shell script that simply cannot be written safely.


> For instance, in the case of source-maintenance.sh I will address the proper quoting of variable names in a followup PR instead of taking on it here

If we are talking about old/unchanged code, then, yes, please do _not_ fix its quoting in this PR. If we are talking about new/changed code, then it is best to quote properly (by default) unless there are very compelling reasons not to do it. The same approach applies to NULL, HERE, const, auto, and many other legacy code issues we have to deal with every day. Nothing special here.

 

> > > ... testing by adding Wextra in test builds
> > 
> > 
> > Probably not -- `-Wextra` tests would fail (until the very last PR in the series). Instead, each "class of errors" PR would simply add `-Wfoo` option(s) that represent that specific class. Those tests would succeed if the PR covered all the cases in the PR-specific class.
 
> make -k and search in the logs for an error message would enable iterations just fine, wouldn't they?

The tricks you use while working on the changes in your workspace are up to you, but for an "enforce `-Wfoo`" PR to be officially merged, the official tests have to succeed and official testing for `-Wfoo` should be enabled, so that we can check that all cases got covered, and so that all future PRs have to maintain the newly achieved status quo. Adding `-Wextra` is not an option for such an incremental approach.

Similarly, a "fix a few `-Wfoo` problems I could find in my manual tests" is not a good PR scope, while "fix all `-Wfoo` problems" is very good (the "... that we can find" condition is always assumed, of course).
 
 
> I get your point on them staying fixed.

Yes, auto-enforced invariants are much better than essentially temporary fixes. There is also a point about _all_ of those `-Wfoo` issues being uncovered (to the extent our CI can uncover them). A developer using tricks can easily miss a case that CI checks. We see such missed cases all the time in PRs, and that if fine -- that is what CI is for.

 
 
 
> > I do not think we should have external variables for ASTYLE and CHECKSUM features. That disagreement remains unresolved.
 
> I believe it follows best practices. Besides my previous example on ./configure (think CC, CXXFLAGS, etc), it is also used in most Unix software. A software using an environment variable to pass command line options is a white fly - I can't think of a single example. Whereas EDITOR is a de-facto standard, for example.

There is a big difference between the standard `CC` or `EDITOR` and, say, `CLANG` or `EMACS`. You are proposing the latter. I would not have objected to using the former (if it is actually needed).

As for `CXXFLAGS`, it is for a C++ compiler, what `SQUID_SOURCE_MAINTENANCE_OPTIONS` essentially is for our source maintenance script. I did not invent that solution, of course. There are many similar examples: `less` has `LESS` (for options, not the executable location!) and git-diff has `GIT_DIFF_OPTS`.

There are, of course, many examples of programs using custom environment variables that duplicate command line options and even some that use custom environment variables that control some functionality that cannot be controlled via a command line option. However, most of those examples either solve a more complicated problem (e.g., several related programs need the same set of controls) or are just examples of poor engineering (no apparent consistency in what is controlled by what -- each piece is scratching some specific itch with admin wasting time, constantly looking for the right solution to each new problem). We do not need the former and should avoid the latter.


 
> > BTW, why do you need `MD5`/`CHECKSUM` at all? What is the actual use case there, assuming the script is enhanced to find/guess a usable checksum program? Is your favorite checksum program not on the proposed guessing list?
 
> It is.

Case closed for `MD5`/`CHECKSUM` then, I think: We should not add unnecessary external configuration vectors. Your use case (and many more!) is already addressed nicely by your checksum guessing changes.


> > > * I'll fix the debug-sections bug (by not updating debug-sections.txt if we don't scan all files)

> > This is a tricky problem: We should discuss whether it would be better to keep debug-sections.txt in sync (by scanning all files) instead of creating a special case and leaving debug-sections.txt stale (by not updating it at all). What do you think?
 
> a full source scan would kill feasibility of the (arguably way more common) use case of a pre-commit hook to enforce source formatting on changed parts only, to make it fast

There may be a good use case budding here, _but_ it probably needs a lot more work, and that work may prove to be too much:

1. I very much doubt that it is a good idea for a pre-commit hook to modify manually-written source code files, for several reasons: automatically modifying sources is dangerous -- astyle can break files in irrecoverable way and our current git protections would be bypassed by this kind of hook; some unformatted commits are actually needed; a formatting hook would make it very difficult to commit just the _staged_ changes (another very common case!). I suspect that, if hooks are easy to bypass, then this hook should just reject commits with unformatted files, leaving the admin with an option to either format or force a hookless commit. To build such a hook, we would need to add support for _detection_ of unformatted files (without modifying anything) -- another script configuration option (similar to a well-known `--dry-run` option that many tools use).

2. We should disable other slow/expensive/global/destructive actions as well then. I am thinking gperf, but there may be more.


BTW, I bet we can make debug-sections construction fast. We already have checkMakeNamedErrorDetails which is global but fast. I bet we can rebuild debug sections in under a second using a similar approach. This improvement can be a part of this or a dedicated PR. If it is delayed to a dedicated PR, then this PR should ignore debug-sections speed.

Reply:
> > TBH, this script is a right mess.
> 
> Yes, it is. And so is a lot of Squid code.
> 
> > I wouldn't be surprised if it broke if any file name contained a space.
> 
> Yes, I bet it will break.
> 
> > I will address that in a followup PR
> 
> FWIW, we do not want to support spaces in git-controlled paths. However, "external" paths (computed or supplied by admins) may indeed contain spaces, especially on Windows, and we should support them.

Agreed.

> I suspect we should switch from shell to Perl (or better) instead of spending a lot of time on perfecting a non-trivial shell script that simply cannot be written safely.

This script may not be trivial but it's still simple enough, and plays well to the shell's strenths

> > For instance, in the case of source-maintenance.sh I will address the proper quoting of variable names in a followup PR instead of taking on it here
> 
> If we are talking about old/unchanged code, then, yes, please do _not_ fix its quoting in this PR. If we are talking about new/changed code, then it is best to quote properly (by default) unless there are very compelling reasons not to do it. The same approach applies to NULL, HERE, const, auto, and many other legacy code issues we have to deal with every day. Nothing special here.

I agree.
I would however, for this specific script, take a look at quoting in a followup PR. There's two options: one is to add minimal changes to make it safe and readable, and the other one would be to enforce some kind of standard (e.g. use CamelCase or whatever style we converge on, and use ${} everywhere. It is one fewer decision points, and ${bare} is a functional superset of $bare

> > > > ... testing by adding Wextra in test builds
> > > 
> > > 
> > > Probably not -- `-Wextra` tests would fail (until the very last PR in the series). Instead, each "class of errors" PR would simply add `-Wfoo` option(s) that represent that specific class. Those tests would succeed if the PR covered all the cases in the PR-specific class.
> 
> > make -k and search in the logs for an error message would enable iterations just fine, wouldn't they?
> 
> The tricks you use while working on the changes in your workspace are up to you, but for an "enforce `-Wfoo`" PR to be officially merged, the official tests have to succeed and official testing for `-Wfoo` should be enabled, so that we can check that all cases got covered, and so that all future PRs have to maintain the newly achieved status quo. Adding `-Wextra` is not an option for such an incremental approach.
> 
> Similarly, a "fix a few `-Wfoo` problems I could find in my manual tests" is not a good PR scope, while "fix all `-Wfoo` problems" is very good (the "... that we can find" condition is always assumed, of course).
> 
> > I get your point on them staying fixed.
> 
> Yes, auto-enforced invariants are much better than essentially temporary fixes. There is also a point about _all_ of those `-Wfoo` issues being uncovered (to the extent our CI can uncover them). A developer using tricks can easily miss a case that CI checks. We see such missed cases all the time in PRs, and that if fine -- that is what CI is for.
> 
> > > I do not think we should have external variables for ASTYLE and CHECKSUM features. That disagreement remains unresolved.
> 
> > I believe it follows best practices. Besides my previous example on ./configure (think CC, CXXFLAGS, etc), it is also used in most Unix software. A software using an environment variable to pass command line options is a white fly - I can't think of a single example. Whereas EDITOR is a de-facto standard, for example.
> 
> There is a big difference between the standard `CC` or `EDITOR` and, say, `CLANG` or `EMACS`. You are proposing the latter. I would not have objected to using the former (if it is actually needed).

I fail to see the difference between EMACS and ASTYLE. CHECKSUM is less compelling and I'll be happy to remove it if it helps go forward

> As for `CXXFLAGS`, it is for a C++ compiler, what `SQUID_SOURCE_MAINTENANCE_OPTIONS` essentially is for our source maintenance script. I did not invent that solution, of course. There are many similar examples: `less` has `LESS` (for options, not the executable location!) and git-diff has `GIT_DIFF_OPTS`.
> 
> There are, of course, many examples of programs using custom environment variables that duplicate command line options and even some that use custom environment variables that control some functionality that cannot be controlled via a command line option. However, most of those examples either solve a more complicated problem (e.g., several related programs need the same set of controls) or are just examples of poor engineering (no apparent consistency in what is controlled by what -- each piece is scratching some specific itch with admin wasting time, constantly looking for the right solution to each new problem). We do not need the former and should avoid the latter.

It seems we are at a bit of a stalemate here, we both have working examples of what works.
How about asking someone else to weigh in? e.g. @yadij , would you be willing to express your preference here?


> > > BTW, why do you need `MD5`/`CHECKSUM` at all? What is the actual use case there, assuming the script is enhanced to find/guess a usable checksum program? Is your favorite checksum program not on the proposed guessing list?
> 
> > It is.
> 
> Case closed for `MD5`/`CHECKSUM` then, I think: We should not add unnecessary external configuration vectors. Your use case (and many more!) is already addressed nicely by your checksum guessing changes.

Ok, I'll remove this. I'm much less happy about changing ASTYLE, for two reasons. One, the fact that we rely on a very old version of astyle makes it a much more common use case. Two, the code complexity to support a "default options" environment variable is much higher than the current approach. I'll implement if others chime in and a consensus is built in that direction.

> > > > * I'll fix the debug-sections bug (by not updating debug-sections.txt if we don't scan all files)
> 
> > > This is a tricky problem: We should discuss whether it would be better to keep debug-sections.txt in sync (by scanning all files) instead of creating a special case and leaving debug-sections.txt stale (by not updating it at all). What do you think?
> 
> > a full source scan would kill feasibility of the (arguably way more common) use case of a pre-commit hook to enforce source formatting on changed parts only, to make it fast
> 
> There may be a good use case budding here, _but_ it probably needs a lot more work, and that work may prove to be too much:
> 
> 1. I very much doubt that it is a good idea for a pre-commit hook to modify manually-written source code files, for several reasons: automatically modifying sources is dangerous -- astyle can break files in irrecoverable way and our current git protections would be bypassed by this kind of hook; some unformatted commits are actually needed; a formatting hook would make it very difficult to commit just the _staged_ changes (another very common case!). I suspect that, if hooks are easy to bypass, then this hook should just reject commits with unformatted files, leaving the admin with an option to either format or force a hookless commit. To build such a hook, we would need to add support for _detection_ of unformatted files (without modifying anything) -- another script configuration option (similar to a well-known `--dry-run` option that many tools use).

Good point, but the case for having a fast option for the common case remains

> 2. We should disable other slow/expensive/global/destructive actions as well then. I am thinking gperf, but there may be more.

Yes

> BTW, I bet we can make debug-sections construction fast. We already have checkMakeNamedErrorDetails which is global but fast. I bet we can rebuild debug sections in under a second using a similar approach. This improvement can be a part of this or a dedicated PR. If it is delayed to a dedicated PR, then this PR should ignore debug-sections speed.

I would not address that in this PR; as you say, the path to making this more forceful is long

Reply:
> I would however, for this specific script, take a look at quoting in a followup PR.

No objections from _me_. Just add a note to the PR description or code so that it is clear that you moved that necessary work into a TODO.

> There's two options: one is to add minimal changes to make it safe and readable, and the other one would be to enforce some kind of standard (e.g. use CamelCase or whatever style we converge on, and use ${} everywhere. It is one fewer decision points, and ${bare} is a functional superset of $bare

I do not see these two options as mutually exclusive and exhaustive. I recommend discussing (somewhere) what to do in those future dedicated PR(s) before actually doing it. For example, I do not think we should write `${x}` in contexts where the curly braces bring no value and only raise reader questions/doubts. Its unnecessary, unnatural, and less readable. However, if others insist on requiring curly braces everywhere, I may not have enough time to fight about it.


> > There is a big difference between the standard `CC` or `EDITOR` and, say, `CLANG` or `EMACS`. You are proposing the latter. I would not have objected to using the former (if it is actually needed).
 
> I fail to see the difference between EMACS and ASTYLE. 

There is not. Both are equally bad. I will rephrase in hope to clarify:

There is a big difference between the standard `CC` and a fictional `CLANG`. There is a big difference between the standard `EDITOR` and a fictional `EMACS`. The standard `CC` and `EDITOR` variables help any program choose the right implementation of a common general-purpose tool (i.e. a "C compiler" or an "editor"). The fictional `CLANG` and `EMACS` as well as the proposed `ASTYLE` specify the location of a specific program binary. Unix has better, standard means for locating binaries of known programs.


> How about asking someone else to weigh in? e.g. @yadij , would you be willing to express your preference here?

IMHO, every GitHub comment is an invitation for others to weigh in. I certainly welcome alternative decision making criteria, but I hope it does not come to counting votes over personal preferences. My objection is not based on personal preferences.

 
> I'm much less happy about changing ASTYLE, for two reasons. One, the fact that we rely on a very old version of astyle makes it a much more common use case. 

I bet it is increasingly uncommon for folks to install newer versions of astyle -- the tool is dying so most folks will not have it installed at all. And there are already ways to deal with this unpleasant problem, including an existing `--with-astyle` command line option that any developer would know how to use. And if newer astyle would address this pain, then we should just upgrade! I see no justification for adding a yet another configuration vector to address this specific problem, especially when that extra configuration vector is rather non-standard and needs additional work (compared to what is already in this PR).


> Two, the code complexity to support a "default options" environment variable is much higher than the current approach. 

I am not a shell expert, but I bet supporting `SQUID_SOURCE_MAINTENANCE_OPTIONS` is not _difficult_, and we should keep in mind that supporting an environment variable capable of carrying all configuration options is also a much more powerful (and fairly common) instrument.

> I'll implement if others chime in and a consensus is built in that direction.

And I can implement it if it helps to tilt the scales.


> the case for having a fast option for the common case remains

I agree. What bothers me in this area is that we are being mudding or overloading (for the lack of more accurate words) the meaning of `--only-changed-since`. I would expect such an option to ignore unchanged files in _all_ respects and _not_ ignore changed files in any respects, but we are discovering problems with both natural assumptions, and adding exceptions/surprises, sometimes at the expense of much better solutions such as making debug-sections generation much faster.

I suspect we should

* either rename `--only-changed-since` to `--format-changed-since` (in a possibly futile attempt to emphasize that the option does not affect non-formatting code/behavior) _and_ add `--format-only` or `--generate no`
* or replace `--only-changed-since` with `--quick-update-since` (or something like that) and then hard-code all the formatting/generation decisions specifically based on their perceived speed. This is similar to the well-known "quick disk format" -- a quick and dirty solution that should not be mistaken for all the shades of "proper" (slow) formatting.

Reply:
Please use command line options. Safe use of environment variables have a minefield of security issues we don't have time to wade through properly every time the script gets updated or used in a new OS environment.

Reply:
Okay, I will go for command line options.
How about adding a detection for astyle-$TargetAstyleVersion ? It would solve my use case and be nothing different from what we do for CHECKSUM

Reply:
> How about adding a detection for astyle-$TargetAstyleVersion? 

I would not object to it if some distributions actually install `astyle-version` binaries. If no distribution does that, then we would be hard-coding strange new logic/interface instead of reusing one of the already available typical interfaces.

IIRC, source-based astyle installation does not add suffixes to astyle binaries (by default). I know I install mine as `/home/alex/local/astyle-v2.04/bin/astyle`.

> It would solve my use case 

AFAICT, your use case can be solved in many other ways using existing interfaces. Perhaps I am missing something, but I do not understand your persistent resistance to using one of them.

> and be nothing different from what we do for CHECKSUM

There is a difference between finding a program that does X and finding version V of a given program. However, I believe it is OK to add code that finds the right astyle version if distros use version-specific names for some astyle installations. I think it is OK to add such code _not_ because it would be similar to finding a checksum program (conceptually, it would not be similar), but because you apparently want it very much (and willing to do the necessary legwork) _and_ it would help developers using those distributions to run the script without any parameters or tricks -- install astyle-V and go.

Reply:
AFAIK Gentoo, Arch and maybe some BSD flavours use that style of naming to install multiple versions of packages.

So yes it seems reasonable to allow for that when searching for the tool(s).